Blacken salt (Implements SEP 15)

waynew opened the pull request:
### What does this PR do?

What it says on the tin. Implements [SEP 15](https://github.com/saltstack/salt-enhancement-proposals/blob/master/accepted/0015-black.md) - at least the `black` portion of it. Apparently `isort` will cause a bunch of issues with our linting, because of the number of lint comments around imports.

### What issues does this PR fix or reference?

[SEP 15](https://github.com/saltstack/salt-enhancement-proposals/blob/master/accepted/0015-black.md)

### Previous Behavior

Salt's code was painted many different colors of bikeshed

### New Behavior

Salt's code is now a lovely shade of Black

### Tests written?

No (code should not have been changed)

### Commits signed with GPG?

Yes

rares-pop commented:
This is pretty cool! There seem to be some config issues, or really black issues (https://github.com/PyCQA/pycodestyle/issues/373), but things look pretty good!

s0undt3ch reviewed (changes requested):
In the pre-commit config dotfile on Salt's repo, let's add something like:
```yaml
- repo: https://github.com/ambv/black
  rev: stable
  hooks:
  - id: black
    args: [--line-length, '120', .]
```
Ajust `args` to those you used when blackening Salt.

waynew commented:
Well, I'm sure there will be a few more merge conflicts, but this should be much further down the path. Also I fixed all the lint issues that I had, and added black to pre-commit 🎉

s0undt3ch commented on .testing.pylintrc:
```diff
@@ -60,7 +60,7 @@ extension-pkg-whitelist=
 # Fileperms Lint Plugin Settings
 fileperms-default=0644
-fileperms-ignore-paths=setup.py,tests/runtests.py,tests/jenkins*.py,tests/saltsh.py,tests/buildpackage.py
+fileperms-ignore-paths=setup.py,tests/runtests.py,tests/jenkins*.py,tests/saltsh.py,tests/buildpackage.py,tests/unit/files/rosters/ansible/roster.py
```
Why does this last file have wrong perms? Could we just fix the perms instead of ignoring it?

s0undt3ch replied:
Also, if you pulled in the latest master branch changes `.testing.pylintrc` is now gone.

waynew replied:
It's part of the kitchen tutorial, so as far as I'm aware it has to have that permissions.

s0undt3ch replied:
ok. the `.testing.pylintrc` issue is still valid though

s0undt3ch reviewed (changes requested):
Also, any files under `salt/ext` should be ignored by black. These are vendored modules which we should keep as intact as possible.

waynew committed:
Add black and isort to precommit

E203 is *not* pep8 compliant, pylint.

mattp- commented:
is it worth holding off on this until the PR queue from other branches has been considered 'resolved' and handled?

waynew commented:
@mattp- I would say no. Rebasing and blackening the other PRs should avoid merge conflicts there. Also, I'm pretty sure that most or  all the PRs that had tests have been merged. 

Meanwhile, we'll have any new development unable to take advantage of black.

If there was a weeks worth of effort or even maybe a month I might feel otherwise, but there's still quite a lot of work to do in the master port effort. We're going to have to deal with merge conflicts and other issues anyway - at least if the code is blackened first we can avoid any code formatting issues there.

Commit:
Blacken salt

s0undt3ch commented on noxfile.py:
```diff
@@ -1,45 +1,51 @@
+    )
+    sys.stderr.flush()
+    exit(1)
+
```
The idea was to write to stderr even before trying to import nox.
We can wrap this code block with [fmt: on/off](https://black.readthedocs.io/en/stable/the_black_code_style.html#the-black-code-style).

s0undt3ch replied:
Can we move this back up?

s0undt3ch commented on pkg/rpm/build.py:
```diff
@@ -1,50 +1,56 @@
-parser.add_argument('buildid',
-        help='The build id to use i.e. the bit after the salt version in the package name',
-        )
+import salt.version
```
This `salt.version` import should probably go back to where it was, see the `sys.path.append` before it?

s0undt3ch commented on salt/_logging/impl.py:
```diff
@@ -1,17 +1,31 @@
+from salt._logging.mixins import LoggingMixinMeta, NewStyleClassMixin
+from salt.exceptions import LoggingRuntimeError
+from salt.utils.ctx import RequestContext
+from salt.utils.textformat import TextFormat
```
These imports need to go back to where they were. See the comment above where we define the new log levels.

s0undt3ch reviewed (changes requested):
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

Of course I didn't review all files, yet, but there are already some things to address.

waynew commented on .pre-commit-config.yaml:
```diff
@@ -407,36 +392,38 @@ repos:
+    rev: stable
+    hooks:
+      - id: black
+        language_version: python3.7
```
We are - I went ahead and combined them, rather than having to have multiple commits.

Go big or go home, right? :upside_down_face:

s0undt3ch replied:
Sure, but there are still other changes besides `black` and `isort`, those should go.

s0undt3ch replied:
Extra changes removed. Resolving conversation.

waynew committed:
Add Blacken Salt to git-blame-ignore-revs

s0undt3ch commented:
Restarted CentOS 8, if that build passes, let's merge.
We can address my review comments at a latter stage.
Let's not consider them blockers.

waynew commented:
@s0undt3ch https://jenkinsci.saltstack.com/job/pr-centos8-py3/job/PR-55765/15/ \o/

Codecov is complaining, and uh... I think I broke it 🙃 Tried to go to codecov to dismiss that change. but it was very unhappy 😝

s0undt3ch commented:
Ignore codecov

s0undt3ch commented:
🎆

waynew commented:
🖤
